scene/main/node.cpp

```diff
@@ -2513,14 +2517,36 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 
 			Variant value = N->get()->get(name).duplicate(true);
 
-			if (E.usage & PROPERTY_USAGE_ALWAYS_DUPLICATE) {
-				Resource *res = Object::cast_to<Resource>(value);
-				if (res) { // Duplicate only if it's a resource
-					current_node->set(name, res->duplicate());
+			Ref<Resource> res = value;
+			if (res.is_valid()) { // Duplicate only if it's a resource.
+				if (instantiated && res->is_local_to_scene()) {
+					// Find the shared copy of the source resource.
+					HashMap<Ref<Resource>, Ref<Resource>>::Iterator R = resources_local_to_sub_scene.find(res);
+					if (R) {
+						value = R->value;
+					} else {
+						Ref<Resource> current_res = current_node->get(name);
+						if (current_res.is_valid() && current_res->is_local_to_scene()) {
+							// Simply copy the data from the source resource to update the resource that was previously set.
+							current_res->copy_from(res, resources_local_to_sub_scene, true);
+							resources_local_to_sub_scene[res] = current_res;
+							value = current_res;
+						} else { // A copy of the source resource is required to overwrite the previous one.
+							Ref<Resource> dupe = res->duplicate_for_local_scene(node, resources_local_to_sub_scene);
+							resources_local_to_sub_scene[res] = dupe;
+							value = dupe;
+						}
+					}
+				} else if (E.usage & PROPERTY_USAGE_ALWAYS_DUPLICATE) {
+					value = res->duplicate();
 				}
+			}
```

Comment:
It's the third time you are writing this code (first two were in #64487). Can it be moved to some method?